crates/contract/src/call.rs

```diff
@@ -0,0 +1,404 @@
+///
+/// Using [`sol!`][sol]:
+///
+/// ```no_run
+/// # async fn test<P: alloy_contract::private::Provider>(provider: P) -> Result<(), Box<dyn std::error::Error>> {
+/// use alloy_contract::SolCallBuilder;
+/// use alloy_primitives::{Address, U256};
+/// use alloy_sol_types::sol;
+///
+/// sol! {
+///     #[sol(rpc)] // <-- Important!
+///     contract MyContract {
+///         function doStuff(uint a, bool b) public returns(address c, bytes32 d);
+///     }
+/// }
+///
+/// # stringify!(
+/// let provider = ...;
+/// # );
+/// let address = Address::ZERO;
+/// let contract = MyContract::new(address, &provider);
+///
+/// // Through `contract.<function_name>(args...)`
+/// let a = U256::ZERO;
+/// let b = true;
+/// let builder: SolCallBuilder<_, MyContract::doStuffCall> = contract.doStuff(a, b);
+/// let MyContract::doStuffReturn { c: _, d: _ } = builder.call().await?;
+///
+/// // Through `contract.call_builder(&<FunctionCall { args... }>)`:
+/// // (note that this is discouraged because it's inherently less type-safe)
+/// let call = MyContract::doStuffCall { a, b };
+/// let builder: SolCallBuilder<_, MyContract::doStuffCall> = contract.call_builder(&call);
+/// let MyContract::doStuffReturn { c: _, d: _ } = builder.call().await?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Using [`ContractInstance`](crate::ContractInstance):
+///
+/// ```no_run
+/// # async fn test<P: alloy_contract::private::Provider>(provider: P, dynamic_abi: alloy_json_abi::JsonAbi) -> Result<(), Box<dyn std::error::Error>> {
+/// use alloy_primitives::{Address, Bytes, U256};
+/// use alloy_dyn_abi::DynSolValue;
+/// use alloy_contract::{CallBuilder, ContractInstance, DynCallBuilder, Interface, RawCallBuilder};
+///
+/// # stringify!(
+/// let dynamic_abi: JsonAbi = ...;
+/// # );
+/// let interface = Interface::new(dynamic_abi);
+///
+/// # stringify!(
+/// let provider = ...;
+/// # );
+/// let address = Address::ZERO;
+/// let contract: ContractInstance<_> = interface.connect(address, &provider);
+///
+/// // Build and call the function:
+/// let call_builder: DynCallBuilder<_> = contract.function("doStuff", &[U256::ZERO.into(), true.into()])?;
+/// let result: Vec<DynSolValue> = call_builder.call().await?;
+///
+/// // You can also decode the output manually. Get the raw bytes:
+/// let raw_result: Bytes = call_builder.call_raw().await?;
+/// // Or, equivalently:
+/// let raw_builder: RawCallBuilder<_> = call_builder.clone().clear_decoder();
+/// let raw_result: Bytes = raw_builder.call().await?;
+/// // Decode the raw bytes:
+/// let decoded_result: Vec<DynSolValue> = call_builder.decode_output(raw_result, false)?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// [sol]: alloy_sol_types::sol
+#[derive(Clone)]
+#[must_use = "call builders do nothing unless you `.call`, `.send`, or `.await` them"]
+pub struct CallBuilder<P, D> {
+    // TODO: this will not work with `send_transaction` and does not differentiate between EIP-1559
+    // and legacy tx
+    request: CallRequest,
```

Comment:
right for signing we need another step where the request is converted into a TypedTransaction first

Reply:
Related `TypedTransaction` work: https://github.com/alloy-rs/alloy/pull/183